lib/elixir/lib/system.ex

```diff
@@ -409,15 +424,78 @@ defmodule System do
 
   The function must receive the exit status code as an argument.
 
-  If the VM terminates programmatically, via `System.stop/1` or `System.halt/1`,
-  the `at_exit/1` callbacks are not executed.
+  If the VM terminates programmatically, via `System.stop/1`, `System.halt/1`,
+  or exit signals, the `at_exit/1` callbacks are not executed.
   """
   @spec at_exit((non_neg_integer -> any)) :: :ok
   def at_exit(fun) when is_function(fun, 1) do
     :elixir_config.update(:at_exit, &[fun | &1])
     :ok
   end
 
+  defmodule SignalHandler do
+    @moduledoc false
+    @behaviour :gen_event
+
+    @impl true
+    def init({event, fun}) do
+      {:ok, {event, fun}}
+    end
+
+    @impl true
+    def handle_call(_message, state) do
+      {:ok, :ok, state}
+    end
+
+    @impl true
+    def handle_event(signal, {event, fun}) do
+      if signal == event, do: :ok = fun.()
+      {:ok, {event, fun}}
+    end
+
+    @impl true
+    def handle_info(_, {event, fun}) do
+      {:ok, {event, fun}}
+    end
+  end
+
+  @doc """
+  Executes the given `fun`, identified by `id`, whenever
+  there is a `signal`.
+
+  The `id` is used exclusively to remove the signal. The
+  same `id` can be given multiple times and each addition
+  has to be removed explicitly (and in the order
+
+  Each call to this function prepends the given `fun` to
+  execute whenever `signal` is invoked. This means that
+  registering a callback that runs on `:sigquit` won't stop
+  the system from aborting. The `fun` must return `:ok`.
+
+  For lower level control over signals, see `:os.set_signal/2`.
```

Comment:
IIRC only sigusr1, sigquit and sigterm are handled by default. the rest goes to default OS handler. It is necessary to call `:os.set_signal` for every one outside those 3

Reply:
You are right. I will have to redesign this API. I will open up a new PR soon.